Content.Server/Shuttles/DockingSystem.cs

```diff
@@ -0,0 +1,463 @@
+
+        private DockingComponent? GetDockable(PhysicsComponent body, TransformComponent dockingXform)
+        {
+            // Did you know Saltern is the most dockable station?
+
+            // Assume the docking port itself (and its body) is valid
+
+            if (!_mapManager.TryGetGrid(dockingXform.GridID, out var grid) ||
+                !EntityManager.HasComponent<ShuttleComponent>(grid.GridEntityId)) return null;
+
+            var transform = body.GetTransform();
+            var dockingFixture = body.GetFixture(DockingFixture)!;
+
+            Box2? aabb = null;
+
+            for (var i = 0; i < dockingFixture.Shape.ChildCount; i++)
+            {
+                aabb = aabb?.Union(dockingFixture.Shape.ComputeAABB(transform, i)) ?? dockingFixture.Shape.ComputeAABB(transform, i);
+            }
+
+            if (aabb == null) return null;
+
+            var enlargedAABB = aabb.Value.Enlarged(DockingRadius * 1.5f);
+
+            // Get any docking ports in range on other grids.
+            _mapManager.FindGridsIntersectingEnumerator(dockingXform.MapID, enlargedAABB, out var enumerator);
+
+            while (enumerator.MoveNext(out var otherGrid))
+            {
+                if (otherGrid.Index == dockingXform.GridID) continue;
+
+                foreach (var ent in otherGrid.GetAnchoredEntities(enlargedAABB))
+                {
+                    if (!EntityManager.TryGetComponent(ent, out DockingComponent? otherDocking) ||
+                        !otherDocking.Enabled ||
+                        !EntityManager.TryGetComponent(ent, out PhysicsComponent? otherBody)) continue;
+
+                    var otherTransform = otherBody.GetTransform();
+                    var otherDockingFixture = otherBody.GetFixture(DockingFixture)!;
+
+                    for (var i = 0; i < otherDockingFixture.Shape.ChildCount; i++)
+                    {
+                        var otherAABB = otherDockingFixture.Shape.ComputeAABB(otherTransform, i);
+
+                        if (!aabb.Value.Intersects(otherAABB)) continue;
+
+                        // TODO: Need CollisionManager's GJK for accurate bounds
+                        // Realistically I want 2 fixtures anyway but I'll deal with that later.
+                        return otherDocking;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private void OnShutdown(EntityUid uid, DockingComponent component, ComponentShutdown args)
+        {
+            if (component.DockJoint == null ||
+                component.DockedWith == null ||
+                EntityManager.GetComponent<MetaDataComponent>(uid).EntityLifeStage > EntityLifeStage.MapInitialized) return;
+
+            Cleanup(component);
+        }
+
+        private void Cleanup(DockingComponent dockA)
+        {
+            _jointSystem.RemoveJoint(dockA.DockJoint!);
+
+            var dockB = dockA.DockedWith!;
+
+            dockA.DockedWith!.DockedWith = null;
```

Comment:
handle 👏 nullables 👏 sanely 👏

Reply:
Do I just nullcheck and throw an exception if it is?